File: 01 Arrays and Strings/Arrays and Strings/1.6-String-Compression/Compress.java
import java.util.*;

public class Compress  {

    public String compress(String str) {
	StringBuilder compressed = new StringBuilder();
	int count = 0;
	
	for (int i = 0; i < str.length(); ++i) {
	    ++count;
	    if  (i == str.length() - 1  || str.charAt(i) != str.charAt(i+1)) {
		compressed.append(str.charAt(i));
		compressed.append(count);
		count = 0;
	    }
	}

	String compstr = compressed.toString();
	return compstr.length() > str.length() ? str : compstr;
    }

    public static void main(String[] args) {
	Compress comp = new Compress();
	System.out.print("Enter a string to compress: ");
	Scanner in = new Scanner(System.in);
	System.out.println(comp.compress(in.next()));
	in.close();
    }
}

File: 01_arrays_and_strings/cpp/compress.cpp
#include <string>
#include <iostream>
#include <sstream>

std::string compress(std::string str) {
  int count = 0;
  std::ostringstream compressed;

  for (auto it = str.begin(); it != str.end(); ++it) {
    ++count;
    /* If the next character is different from the last current
       or we've reached the end of the string */
    if (std::next(it) == str.end() || *std::next(it) != *it) {
      compressed << *it << count;
      count = 0;
    }
  }

  std::string compstr = compressed.str();
  return compstr.size() >= str.size() ? str : compstr;
}

int main() {
  std::cout << "Enter a string to compress: ";
  std::string str;
  std::cin >> str;
  std::cout << compress(str) << std::endl;
  return 0;
}
